File: Sorting Algorithms/Bitonic Sort/BitonicSort.java
// Java program for Bitonic Sort. Note that this program works only when size of input is a power of 2.

import java.util.*;

public class BitonicSort {
    /* The parameter dir indicates the sorting direction, ASCENDING
       or DESCENDING; if (a[i] > a[j]) agrees with the direction,
       then a[i] and a[j] are interchanged. */
    void compAndSwap(int a[], int i, int j, int dir) {
        if ((a[i] > a[j] && dir == 1) || (a[i] < a[j] && dir == 0)) {
            // Swapping elements
            int temp = a[i];
            a[i] = a[j];
            a[j] = temp;
        }
    }

    /* Recursively sorts a bitonic sequence in ascending order, if dir = 1,
       and in descending order otherwise (means dir=0). The sequence to be sorted
       starts at index position low, the parameter cnt is the number of elements to be sorted. */
    void bitonicMerge(int a[], int low, int cnt, int dir) {
        if (cnt > 1) {
            int k = cnt / 2;
            for (int i = low; i < low + k; i++)
                compAndSwap(a, i, i + k, dir);
            bitonicMerge(a, low, k, dir);
            bitonicMerge(a, low + k, k, dir);
        }
    }

    /* This function first produces a bitonic sequence by recursively sorting
       its two halves in opposite sorting orders, and then calls bitonicMerge
       to make them in the same order */
    void bitonicSort(int a[], int low, int cnt, int dir) {
        if (cnt > 1) {
            int k = cnt / 2;
            bitonicSort(a, low, k, 1);
            bitonicSort(a, low + k, k, 0);
            bitonicMerge(a, low, cnt, dir);
        }
    }

    /* A utility function to print array of size n */
    void printArray(int a[]) {
        int n = a.length;
        for (int i = 0; i < n; i++)
            System.out.print(a[i] + " ");
        System.out.println();
    }

    public static void main(String args[]) {
        int a[] = {3, 7, 4, 8, 6, 2, 1, 5};
        int n = a.length;
        int dir = 1; // 1 represents ascending order, 0 represents descending order

        BitonicSort ob = new BitonicSort();
        ob.bitonicSort(a, 0, n, dir);

        System.out.println("Sorted array: ");
        ob.printArray(a);
    }
}

File: Sorting Algorithms/Bitonic Sort/BitonicSort.cs
using System;

class BitonicSort
{
    void BitonicMerge(int[] arr, int low, int count, bool dir)
    {
        if (count > 1)
        {
            int k = count / 2;
            for (int i = low; i < low + k; i++)
                CompareAndSwap(arr, i, i + k, dir);
            BitonicMerge(arr, low, k, dir);
            BitonicMerge(arr, low + k, k, dir);
        }
    }

    void BitonicSortRecursive(int[] arr, int low, int count, bool dir)
    {
        if (count > 1)
        {
            int k = count / 2;
            BitonicSortRecursive(arr, low, k, true);
            BitonicSortRecursive(arr, low + k, k, false);
            BitonicMerge(arr, low, count, dir);
        }
    }

    public void BitonicSort(int[] arr, int count, bool dir)
    {
        BitonicSortRecursive(arr, 0, count, dir);
    }

    void CompareAndSwap(int[] arr, int i, int j, bool dir)
    {
        if ((arr[i] > arr[j] && dir) || (arr[i] < arr[j] && !dir))
        {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }

    public static void Main()
    {
        int[] arr = { 3, 7, 4, 8, 6, 2, 1, 5 };
        int count = arr.Length;
        bool dir = true; // true for ascending, false for descending
        BitonicSort bs = new BitonicSort();
        bs.BitonicSort(arr, count, dir);

        Console.WriteLine("Sorted array:");
        for (int i = 0; i < count; i++)
            Console.Write(arr[i] + " ");
    }
}
